src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/ScreenRecorder.cs

```diff
@@ -100,18 +103,92 @@ private void OnDestroy()
+            });
+            CreateAndUploadGameplaySession(endTime);
+        }
+
+        private async void CreateAndUploadGameplaySession(DateTime endTime)
+        {
+            RGDebug.LogInfo($"Creating and uploading GameplaySession on the backend, from {_startTime} to {endTime} with {_tickNumber} ticks");
+            try
+            {
+                
+                // First, create the gameplay session
+                long gameplaySessionId = -1;
+                await _rgServiceManager.CreateGameplaySession(_startTime, endTime, _tickNumber,
+                    (response) =>
+                    {
+                        gameplaySessionId = response.id;
+                        RGDebug.LogInfo($"Created gameplay session with id: {response.id}");
+                    },
+                    () => { RGDebug.LogError($"Failed to create gameplay session"); });
+                
+                // If the gameplay session was not created, return
+                if (gameplaySessionId == -1)
                 {
-                    RGDebug.LogInfo($"Zipping recording replay to file: {_currentVideoDirectory}.zip");
-                    ZipFile.CreateFromDirectory(theVideoDirectory, theVideoDirectory + ".zip");
-                    Directory.Delete(theVideoDirectory, true);
-                    RGDebug.LogInfo($"Finished zipping replay to file: {_currentVideoDirectory}");
-                });
-                _isRecording = false;
+                    return;
+                }
+                
+                // Upload the gameplay session data
+                await _rgServiceManager.UploadGameplaySessionData(gameplaySessionId,
+                    _currentGameplaySessionDirectoryPrefix + "_data.zip",
+                    () => { RGDebug.LogInfo($"Uploaded gameplay session data"); },
+                    () => { RGDebug.LogError($"Failed to upload gameplay session data"); });
+                
+                // Next, upload the gameplay session screenshots
+                await _rgServiceManager.UploadGameplaySessionScreenshots(gameplaySessionId,
+                    _currentGameplaySessionDirectoryPrefix + "_screenshots.zip",
+                    () => { RGDebug.LogInfo($"Uploaded gameplay session screenshots"); },
+                    () => { RGDebug.LogError($"Failed to upload gameplay session screenshots"); });
+                
+                // Finally, upload the thumbnail
+                await _rgServiceManager.UploadGameplaySessionThumbnail(gameplaySessionId,
+                    _currentGameplaySessionDirectoryPrefix + "_thumbnail.png",
+                    () => { RGDebug.LogInfo($"Uploaded gameplay session thumbnail"); },
+                    () => { RGDebug.LogError($"Failed to upload gameplay session thumbnail"); });
+            }
+            catch (Exception e)
+            {
+                Debug.LogError(e);
             }
+            RGBotManager.GetInstance().ShowUploadingIndicator(false);
```

Comment:
This will ultimately call .`SetActive` off of the main thread in Unity... dragons ahead...

you can either make the function toggle a state variable that is checked in Update() of RGBotManager to enable/disable the indicator on the next update, or you can deal with launching this in a coroutine with `yield return null;` to make it do it on the main thread in the next frame, or some other cleverness to do this action on the next main thread update call